bokeh/server/auth_provider.py

```diff
@@ -0,0 +1,318 @@
+    the module, or None otherwise.
+
+    '''
+
+    def __init__(self, module_path):
+        if not isfile(module_path):
+            raise ValueError()
+
+        self._module = load_auth_module(module_path)
+
+        super(AuthModule, self).__init__()
+
+    @property
+    def get_user(self):
+        return getattr(self._module, 'get_user', None)
+
+    @property
+    def get_user_async(self):
+        return getattr(self._module, 'get_user_async', None)
+
+    @property
+    def login_url(self):
+        return getattr(self._module, 'login_url', None)
+
+    @property
+    def get_login_url(self):
+        return getattr(self._module, 'get_login_url', None)
+
+    @property
+    def login_handler(self):
+        return getattr(self._module, 'LoginHandler', None)
+
+    @property
+    def logout_url(self):
+        return getattr(self._module, 'logout_url', None)
+
+    @property
+    def logout_handler(self):
+        return getattr(self._module, 'LogoutHandler', None)
+
+class NullAuth(AuthProvider):
+    ''' A default no-auth AuthProvider.
+
+    All of the properties of this provider return None.
+
+    '''
+    @property
+    def get_user(self):
+        return None
+
+    @property
+    def get_user_async(self):
+        return None
+
+    @property
+    def login_url(self):
+        return None
+
+    @property
+    def get_login_url(self):
+        return None
+
+    @property
+    def login_handler(self):
+        return None
+
+    @property
+    def logout_url(self):
+        return None
+
+    @property
+    def logout_handler(self):
+        return None
+
+#-----------------------------------------------------------------------------
+# Dev API
+#-----------------------------------------------------------------------------
+
+def load_auth_module(module_path):
+    ''' Load a Python source file at a given path as a module.
+
+    Arguments:
+        module_path (str): path to a Python source file
+
+    Returns
+        module
+
+    '''
+    module_name ="bokeh.auth_" + make_globally_unique_id().replace('-', '')
+    try:
+        import importlib.util
+        spec = importlib.util.spec_from_file_location(module_name, module_path)
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)
+
+    except ImportError:
+        try:
+            # python 3.4
+            from importlib.machinery import SourceFileLoader
+            module = SourceFileLoader(module_name, module_path).load_module()
+
+        except ImportError:
+            # python 2
+            import imp
+            module = imp.load_source(module_name, module_path)
+
+    return module
+
+def probably_relative_url(url):
+    ''' Return True if a URL is not one of the common absolute URL formats.
+
+    Arguments:
+        url (str): a URL string
+
+    Returns
+        bool
+
+    '''
+    return not (url.startswith("http:") or url.startswith("https:") or url.startswith("//"))
```

Comment:
Just realized - could be simplified to `return not url.startswith(("http://", "https://", "//"))`. Even on Python 2.7.

Reply:
https://github.com/bokeh/bokeh/pull/9190/commits/5a76573f27f177d24981570445bee59451c8f43b